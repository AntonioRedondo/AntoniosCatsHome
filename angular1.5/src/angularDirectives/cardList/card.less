@import (reference) "../../style/variables";


.card {
	width: 100%;
	height: 250px;
	margin: 40px 0;
	
	display: flex;	
	justify-content: space-between;
	align-items: center;
	
	background: @grayE;
	box-shadow: 0 -5px 0 @purple2, 0 0 10px rgba(0, 0, 0, 0.5);
	
	color: @gray3;
	
	transition: all @animTime1 ease-out;
	
	@media (max-width: @mobileBreakpoint) {
		height: 150px;
	}
	
	&:hover {
		background: white;
		box-shadow: 0 -5px 0 orange, 10px 10px 30px rgba(0, 0, 0, 0.2);
		cursor: pointer;
		transform: scale(1.025);
	}
	
	&:active {
		transform: scale(1);
		background: lightyellow;
	}
	
	@media (max-width: @mobileBreakpoint) {
		margin: 0;
		
		box-shadow: 0 5px 0 @purple3 inset, 10px 10px 30px rgba(0, 0, 0, 0.2);
	}
	
	&--selected {
		background: papayawhip;
	}
	
	&--selected:hover {
		background: papayawhip;
	}
	
	&__photo {
		width: 250px;
		height: 250px;
		
		position: relative;
		
		@media (max-width: @mobileBreakpoint) {
			width: 150px;
			min-width: 150px;
			height: 150px;
		}
	}
	
	&__photo-img {
		width: 100%;
		height: 100%;
		display: block;
		
		position: absolute;
		top: 0;
		left: 0;
		
		background-size: cover;
	}
	
	&__body {
		padding: 40px;
		overflow: hidden;
		
		flex: 1;
		
		font-size: 1em;
		
		@media (max-width: @mobileBreakpoint) {
			padding-right: 0;
		}
	}
	
	&__body-title {
		margin: 0;
		
		font-family: "Rokkitt", serif;
		font-size: 2.5em;
	}
	
	&__body-text {
		margin: 0 0 20px;
		
		color: @gray6;
		
		@media (max-width: @mobileBreakpoint) {
			display: none;
		}
	}
}
