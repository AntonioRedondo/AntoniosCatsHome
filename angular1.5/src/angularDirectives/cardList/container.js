angular
	.module("ach")

	.directive("cardList", ["$achQueries", function($achQueries) {
		return {
			restrict: "E",
			templateUrl: "angularDirectives/cardList/card.tpl.htm",
			scope: {},
			link: function(scope) {
				$achQueries.cards().then(function(results) {
					scope.cards = results.sort(function(a, b) {
						if (a.name < b.name)
							return -1;
						if (a.name > b.name)
							return 1;
						else return 0;
					});
					scope.cardsOriginal = scope.cards;
				}, function(e) {
					console.debug(e);
				});

				scope.cardSelected = -1;
				scope.cardClicked = function($index) {
					if ($index === scope.cardSelected)
						scope.cardSelected = -1;
					else scope.cardSelected = $index;
				};
				
				scope.filter = function() {
					scope.cards = scope.cardsOriginal.filter(function(item) {
						var itemKeys = Object.keys(item);
						for (var n=0; n<itemKeys.length; ++n)
							if (item[itemKeys[n]] && item[itemKeys[n]].toLowerCase().indexOf(scope.filterString.toLowerCase()) > -1)
								return true;
						return false;
					});
				};
			}
		};
	}]);
