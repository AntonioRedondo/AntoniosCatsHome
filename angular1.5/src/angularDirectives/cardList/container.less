@import (reference) "../../style/variables";


.filter {
	width: 100%;
	max-width: @width;
	margin: 50px auto;
	padding: 8px 20px 10px;
	display: block;
	
	background: white;
	border-radius: 5px;
	border: 4px solid @grayB;
	// box-shadow: 0 0 20px rgba(255, 255, 255, 0), 0 0 10px rgba(0, 0, 0, 0.2) inset;
	
	color: @gray3;
	font-size: 1.8em;
	font-family: "Open Sans", san-serif;
	
	transition: all 0.3s ease-out;
	
	&:hover {
		border-color: orange;
	}
	
	&:focus {
		border-color: @green1;
		box-shadow: 0 0 20px rgba(255, 255, 255, 0.2), 0 0 10px rgba(0, 0, 0, 0.2) inset;
	}
	
	@media (max-width: @mobileBreakpoint) {
		width: calc(~"100% - 100px");
	}
}
