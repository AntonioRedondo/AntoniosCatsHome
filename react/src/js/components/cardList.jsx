import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { TransitionGroup, CSSTransition } from "react-transition-group";

import actionCreators from "../redux/actionCreators";
import { filterItems } from "../redux/selectors";
import Card from "./card.jsx";
import v from "../style/variables";

export class CardList extends Component {
	constructor(props) {
		super();
		
		this.setSelected = props.setSelected.bind(this);
	}
	
	componentDidMount() {
		if (this.props.items.length === 0)
			this.props.requestCatList();
	}
	
	shouldComponentUpdate({ items, itemSelected }) {
		return items.length !== this.props.items.length
				|| itemSelected !== this.props.itemSelected;
	}
	
	_renderCards() {
		return this.props.items.map(item =>
			<CSSTransition
				key={ item.id }
				classNames="card-list"
				timeout={ v.time2 } >
				<Card
					key={ item.id }
					id={ item.id }
					name={ item.name }
					description={ item.description }
					itemSelected={ this.props.itemSelected }
					onClick={ this.props.setSelected } >
				</Card>
			</CSSTransition>
		);
	}
	
	render() {
		return <TransitionGroup>{ this._renderCards() }</TransitionGroup>;
	}
}

CardList.displayName = "CardList";

CardList.propTypes = {
	itemSelected: PropTypes.string,
	items: PropTypes.array.isRequired,
	requestCatList: PropTypes.func.isRequired,
	setSelected: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
	items: filterItems(state),
	itemSelected: state.itemSelected
});
const mapDispatchToProps = dispatch => ({
	requestCatList: () => dispatch(actionCreators.requestCatList()),
	setSelected: id => () => dispatch(actionCreators.setSelected(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(CardList);
