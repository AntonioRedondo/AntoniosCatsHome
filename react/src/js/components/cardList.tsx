import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { TransitionGroup, CSSTransition } from "react-transition-group";

import actionCreators from "../redux/actionCreators";
import { filterCats } from "../redux/selectors";

import Card from "./card.jsx";
import v from "../style/variables";

export class CardList extends Component {
	static propTypes = {
		catSelected: PropTypes.string,
		cats: PropTypes.array.isRequired,
		requestCats: PropTypes.func.isRequired,
		setSelected: PropTypes.func.isRequired,
	}
	
	componentDidMount() {
		!this.props.cats.length && this.props.requestCats();
	}
	
	shouldComponentUpdate({ cats, catSelected }) {
		return cats.length !== this.props.cats.length
				|| catSelected !== this.props.catSelected;
	}
	
	private renderCards = () => {
		return this.props.cats.map(cat =>
			<CSSTransition
				key={ cat.id }
				classNames="card-list"
				timeout={ v.time2 } >
				<Card
					key={ cat.id }
					id={ cat.id }
					name={ cat.name }
					description={ cat.description }
					catSelected={ this.props.catSelected }
					onClick={ this.props.setSelected } >
				</Card>
			</CSSTransition>
		);
	}
	
	render() {
		return <TransitionGroup>{ this.renderCards() }</TransitionGroup>;
	}
}

const mapStateToProps = state => ({
	cats: filterCats(state),
	catSelected: state.catSelected
});
const mapDispatchToProps = dispatch => ({
	requestCats: () => dispatch(actionCreators.requestCats()),
	setSelected: id => () => dispatch(actionCreators.setSelected(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(CardList);
