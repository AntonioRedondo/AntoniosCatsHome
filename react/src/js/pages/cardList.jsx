import React from "react";
import PropTypes from "prop-types";
import { CSSTransitionGroup } from "react-transition-group";
import { connect } from "react-redux";

import actions from "../redux/actions";
import { filterItems } from "../redux/selectors";
import Card from "../components/card.jsx";
import CardFilter from "../components/cardFilter.jsx";

export class CardList extends React.Component {
	constructor(props) {
		super();
		
		this.setSelectedBound = props.setSelected.bind(this);
		this.setFilterStringBound = props.setFilterString.bind(this);
	}
	
	componentDidMount() {
		if (this.props.items.length === 0)
			this.props.requestItems();
	}
	
	render() {
		return (
			<div>
				<CardFilter onChange={ this.setFilterStringBound } value={ this.props.filterString } />
				<CSSTransitionGroup
					transitionName="card-list"
					transitionEnterTimeout={ 250 }
					transitionLeaveTimeout={ 250 } >
					{
						this.props.items.map(item =>
							<Card
								key={ item.id }
								id={ item.id }
								name={ item.name }
								description={ item.description }
								itemSelected={ this.props.itemSelected }
								onClick={ this.setSelectedBound } >
							</Card>
						)
					}
				</CSSTransitionGroup>
			</div>
		);
	}
}

CardList.displayName = "CardList";

CardList.propTypes = {
	filterString: PropTypes.string,
	itemSelected: PropTypes.string,
	items: PropTypes.array,
	requestItems: PropTypes.func,
	setFilterString: PropTypes.func,
	setSelected: PropTypes.func,
};

const mapStateToProps = state => ({
	items: filterItems(state),
	itemSelected: state.itemSelected,
	filterString: state.filterString
});
const mapDispatchToProps = dispatch => ({
	requestItems: () => dispatch(actions.requestItems()),
	setSelected: id => () => dispatch(actions.setSelected(id)),
	setFilterString: event => dispatch(actions.setFilterString(event.target.value))
});

export default connect(mapStateToProps, mapDispatchToProps)(CardList);
