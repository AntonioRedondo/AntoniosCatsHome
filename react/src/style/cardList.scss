.card-list {
	&-enter {
		height: 0;
		margin: 0;
		opacity: 0;
		transition: all $animTime2 ease-out;
	}
	
	&-enter&-enter-active {
		height: 250px;
		margin: 0 0 40px;
		opacity: 1;
	}
	
	&-leave {
		transition: all $animTime2 ease-in;
	}
	
	&-leave&-leave-active {
		height: 0;
		margin: 0;
		opacity: 0;
	}
}
