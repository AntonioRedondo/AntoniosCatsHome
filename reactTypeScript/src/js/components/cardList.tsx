import React, { Component } from "react";
import { connect } from "react-redux";
import { CSSTransition, TransitionGroup } from "react-transition-group";

import actionCreators from "../redux/actionCreators";
import { filterCats } from "../redux/selectors";

import Card, { ICardProps } from "./card";
import v from "../style/variables";

interface ICardListProps {
	catSelected?: string
	cats?: ICardProps[]
	requestCats?: () => void
	setSelected?: (id: string) => any
}

export class CardList extends Component<ICardListProps> {
	componentDidMount() {
		!this.props.cats.length && this.props.requestCats();
	}
	
	shouldComponentUpdate(nextProps) {
		const { cats, catSelected } = nextProps;
		
		return cats.length !== this.props.cats.length || catSelected !== this.props.catSelected;
	}
	
	renderCards = () => {
		return this.props.cats.map(cat =>
			<CSSTransition
				key={ cat.id }
				classNames="card-list"
				timeout={ v.time2 } >
				<Card
					key={ cat.id }
					id={ cat.id }
					name={ cat.name }
					description={ cat.description }
					catSelected={ this.props.catSelected === cat.id }
					onClick={ this.props.setSelected } />
			</CSSTransition>
		);
	}
	
	render() {
		return <TransitionGroup>{ this.renderCards() }</TransitionGroup>;
	}
}

const mapStateToProps = state => ({
	catSelected: state.catSelected,
	cats: filterCats(state)
});
const mapDispatchToProps = dispatch => ({
	requestCats: () => dispatch(actionCreators.requestCats()),
	setSelected: id => () => dispatch(actionCreators.setSelected(id))
});

export default connect<{}, {}, ICardListProps>(mapStateToProps, mapDispatchToProps)(CardList);
