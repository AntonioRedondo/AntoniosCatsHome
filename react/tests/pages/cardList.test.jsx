import React from "react";
import ReactShallowRenderer  from "react-test-renderer/shallow";
import "jest-styled-components";

import { CardList } from "../../src/js/pages/cardList.jsx";
import mockCatList from "../../src/data/cats.json";


describe("Given the CardList component", () => {
	const shallowRenderer = new ReactShallowRenderer ();
	let tree;
	let mockRequiredPropFunction = () => {};

	beforeEach(() => {
		shallowRenderer.render(
			<CardList
				items={ mockCatList }
				setFilterString={ mockRequiredPropFunction }
				setSelected={ mockRequiredPropFunction }  />
		);
		tree = shallowRenderer.getRenderOutput();
	});
	
	describe("When the component is mounted", () => {
		it("Should render children", () => {
			expect(tree).toMatchSnapshot();
		});
	});
	
});
