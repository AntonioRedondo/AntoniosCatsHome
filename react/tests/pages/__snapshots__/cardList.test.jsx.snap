// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the CardList component When the component is mounted Should render children 1`] = `
<div>
  <CardFilter
    onChange={[Function]}
    value={undefined}
  />
  <CSSTransitionGroup
    transitionAppear={false}
    transitionEnter={true}
    transitionEnterTimeout={250}
    transitionLeave={true}
    transitionLeaveTimeout={250}
    transitionName="card-list"
  >
    <Card
      description="Little Sock was brought to Tony's Cats Home as his owners were moving house and couldn't take him with them. He is an affectionate boy that loves nothing more than being stroked and sitting on a lap having extra cuddles, he will purr with delight. "
      id="1"
      itemSelected={undefined}
      name="Little Sock"
      onClick={[Function]}
    />
    <Card
      description="He is a young and active cat. If owners work full-time he will need cat flap access to outside so he can come and go as he pleases throughout the day. Fluffly is a true rescue cat who will make a lovely companion when settled in his new home."
      id="2"
      itemSelected={undefined}
      name="Fluffy"
      onClick={[Function]}
    />
    <Card
      description="Luna is a real sweetie who loves nothing more than a good snuggle up to her favourite people. She is also very playful and loves to play string toy games. She will make a great addition to the family and will provide a lot of love and laughter through the years! "
      id="3"
      itemSelected={undefined}
      name="Luna"
      onClick={[Function]}
    />
    <Card
      description="She is a loving and playful youngster that enjoys interacting with people, being stroked and groomed, she is a cuddly girl. Being young and active Bella will need cat flap access to outside if owners are working full-time, she loves exploring her territory, playing and sunbathing. "
      id="4"
      itemSelected={undefined}
      name="Bella"
      onClick={[Function]}
    />
  </CSSTransitionGroup>
</div>
`;
